sycl/plugins/opencl/pi_opencl.cpp

```diff
@@ -0,0 +1,361 @@
+    *num_devices = 0;
+    result = PI_SUCCESS;
+  }
+  return PI::cast<pi_result>(result);
+}
+
+pi_result OCL(piextDeviceSelectBinary)(
+    pi_device device, // TODO: does this need to be context?
+    pi_device_binary *images, pi_uint32 num_images,
+    pi_device_binary *selected_image) {
+
+  // TODO dummy implementation.
+  // Real implementaion will use the same mechanism OpenCL ICD dispatcher
+  // uses. Something like:
+  //   PI_VALIDATE_HANDLE_RETURN_HANDLE(ctx, PI_INVALID_CONTEXT);
+  //     return context->dispatch->piextDeviceSelectIR(
+  //       ctx, images, num_images, selected_image);
+  // where context->dispatch is set to the dispatch table provided by PI
+  // plugin for platform/device the ctx was created for.
+
+  *selected_image = num_images > 0 ? images[0] : nullptr;
+  return PI_SUCCESS;
+}
+
+pi_result OCL(piQueueCreate)(pi_context context, pi_device device,
+                             pi_queue_properties properties, pi_queue *queue) {
+  assert(queue && "piQueueCreate failed, queue argument is null");
+
+  cl_platform_id curPlatform;
+  cl_int ret_err =
+      clGetDeviceInfo(PI::cast<cl_device_id>(device), CL_DEVICE_PLATFORM,
+                      sizeof(cl_platform_id), &curPlatform, NULL);
+
+  if (ret_err != CL_SUCCESS) {
+    *queue = nullptr;
+    return PI::cast<pi_result>(ret_err);
+  }
+
+  size_t platVerSize;
+  ret_err = clGetPlatformInfo(curPlatform, CL_PLATFORM_VERSION, 0, NULL,
+                              &platVerSize);
+
+  if (ret_err != CL_SUCCESS) {
+    *queue = nullptr;
+    return PI::cast<pi_result>(ret_err);
+  }
+
+  std::string platVer(platVerSize, '\0');
+  ret_err = clGetPlatformInfo(curPlatform, CL_PLATFORM_VERSION, platVerSize,
+                              &platVer.front(), NULL);
+
+  if (ret_err != CL_SUCCESS) {
+    *queue = nullptr;
+    return PI::cast<pi_result>(ret_err);
+  }
+
+  if (platVer.find("OpenCL 1.0") != std::string::npos ||
+      platVer.find("OpenCL 1.1") != std::string::npos ||
+      platVer.find("OpenCL 1.2") != std::string::npos) {
+    *queue = PI::cast<pi_queue>(clCreateCommandQueue(
+        PI::cast<cl_context>(context), PI::cast<cl_device_id>(device),
+        PI::cast<cl_command_queue_properties>(properties), &ret_err));
+    return PI::cast<pi_result>(ret_err);
+  }
+
+  cl_queue_properties CreationFlagProperties[] = {
+      CL_QUEUE_PROPERTIES, PI::cast<cl_command_queue_properties>(properties),
+      0};
+  *queue = PI::cast<pi_queue>(clCreateCommandQueueWithProperties(
+      PI::cast<cl_context>(context), PI::cast<cl_device_id>(device),
+      CreationFlagProperties, &ret_err));
+  return PI::cast<pi_result>(ret_err);
+}
+
+pi_result OCL(piProgramCreate)(pi_context context, const void *il,
+                               size_t length, pi_program *res_program) {
+
+  size_t deviceCount;
+
+  cl_int ret_err = clGetContextInfo(PI::cast<cl_context>(context),
+                                    CL_CONTEXT_DEVICES, 0, NULL, &deviceCount);
```

Comment:
ret_err is not check here, deviceCount can be any number but below is used to initialize a vector.

Reply:
done.